extensions/exchanges/bitfinex/exchange.js

```diff
@@ -0,0 +1,234 @@
+  }
+
+  function retry (method, args) {
+    if (method !== 'getTrades') {
+      console.error(('\nBitfinex API is down! unable to call ' + method + ', retrying in 10s').red)
+    }
+    setTimeout(function () {
+      exchange[method].apply(exchange, args)
+    }, 10000)
+  }
+
+  var orders = {}
+  var exchange = {
+    name: 'bitfinex',
+  //    historyScan: 'backward',
+    makerFee: 0.1,
+
+    getProducts: function () {
+      return require('./products.json')
+    },
+
+    getTrades: function (opts, cb) {
+      var func_args = [].slice.call(arguments)
+      var client = publicClient()
+      var args = joinProduct(opts.product_id)
+      client.trades(args, function (err, body) {
+        if (err) return retry('getTrades', func_args, err)
+        var trades = body.map(function(trade) {
+          return {
+            trade_id: trade.tid,
+            time: n(trade.timestamp).multiply(1000).value(),
+            size: Number(trade.amount),
+            price: Number(trade.price),
+            side: trade.type
+          }
+        })
+        cb(null, trades)
+      })
+    },
+
+    getBalance: function (opts, cb) {
+      var func_args = [].slice.call(arguments)
+      var client = authedClient()
+      client.wallet_balances(function (err, body) {
+        if (err) return retry('getBalance', func_args, err)
+        var balance = {asset: 0, currency: 0}
+        var accounts = _(body).filter(function (body) { return body.type === c.bitfinex.wallet }).forEach(function (account) {
+          if (account.currency.toUpperCase() === opts.currency) {
+            balance.currency = account.amount
+            balance.currency_hold = (account.amount - account.available)
+          }
+          else if (account.currency.toUpperCase() === opts.asset) {
+            balance.asset = account.amount
+            balance.asset_hold = (account.amount - account.available)
+          }
+        })
+        cb(null, balance)
+      })
+    },
+
+    getQuote: function (opts, cb) {
+      var func_args = [].slice.call(arguments)
+      var client = publicClient()
+      var pair = joinProduct(opts.product_id)
+      client.ticker(pair, function (err, body) {
+        if (err) return retry('getQuote', func_args, err)
+        cb(null, { bid : body.bid, ask : body.ask })
+      })
+    },
+
+    cancelOrder: function (opts, cb) {
+      var func_args = [].slice.call(arguments)
+      var client = authedClient()
+      client.cancel_order(opts.order_id, function (err, body) {
+        if (err) return retry('cancelOrder', func_args, err)
+        cb()
+      })
+    },
+
+    buy: function (opts, cb) {
+      var func_args = [].slice.call(arguments)
+      var client = authedClient()
+      if (c.bitfinex.wallet === 'exchange' && typeof opts.type === 'undefined') {
+        opts.type = 'exchange limit'
+      }
+      else if (c.bitfinex.wallet === 'trading' && typeof opts.type === 'undefined') {
+        opts.type = 'limit'
```

Comment:
ref https://github.com/carlos8f/zenbot/pull/233 again, could you add a check for `order_type` (if bitfinex supports it) and use market orders when people supply `taker`? not a must, just a "would be nice"